src/H5Iint.c

```diff
@@ -9925,181 +9471,174 @@ H5I__discard_mt_type_info(H5I_mt_type_info_t * type_info_ptr)
-                fl_stail = atomic_load(&(H5I_mt_g.type_info_fl_stail));
+                    assert(fl_snext.ptr);
 
-                assert(fl_shead.ptr);
-                assert(fl_stail.ptr);
+                    new_fl_shead.ptr = fl_snext.ptr;
+                    new_fl_shead.sn  = fl_shead.sn + 1;
 
-                fl_snext = atomic_load(&(fl_shead.ptr->fl_snext));
+                    if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.type_info_fl_shead), &fl_shead, new_fl_shead) ) {
 
-                test_fl_shead = atomic_load(&(H5I_mt_g.type_info_fl_shead));
+                        /* the attempt to remove the first item from the free list
+                         * failed.  Update stats and try again.
+                         */
+                        atomic_fetch_add(&(H5I_mt_g.num_type_info_fl_head_update_cols), 1ULL);
 
-                if ( ( test_fl_shead.ptr == fl_shead.ptr ) && ( test_fl_shead.sn == fl_shead.sn ) ) {
+                    } else {
 
-                    if ( fl_shead.ptr == fl_stail.ptr ) {
+                        H5I_mt_type_info_sptr_t null_snext = {NULL, 0ULL};
 
-                        if ( NULL == fl_snext.ptr ) {
+                        /* first has been removed from the free list.  Set type_info_ptr to fl_shead.ptr,
+                         * discard *type_info_ptr, update stats, and exit the loop by setting done to true.
+                         */
+                        type_info_ptr = fl_shead.ptr;
 
-                            /* the free list is empty */
-                            atomic_fetch_add(&(H5I_mt_g.num_type_info_fl_frees_skipped_due_to_empty), 1);
-                            done = TRUE;
-                            break;
-                        }
+                        assert(H5I__TYPE_INFO == type_info_ptr->tag);
+                        assert(type_info_ptr->on_fl);
+                        assert(0 < atomic_load(&(type_info_ptr->serial_num)));
 
-                        /* attempt to set H5I_mt_g.type_info_fl_stail to fl_snext.  It doesn't
-                         * matter whether we succeed or fail, as if we fail, it
-                         * just means that some other thread beat us to it.
-                         *
-                         * that said, it doesn't hurt to collect stats
+                        /* the above assert only exists in production builds.  If type_info_ptr->serial_num
+                         * is zero and we get this far, increment H5I_mt_g.num_id_info_fl_head_sn_is_zero.
                          */
-                        new_fl_stail.ptr = fl_snext.ptr;
-                        new_fl_stail.sn  = fl_stail.sn + 1;
-                        if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.type_info_fl_stail), &fl_stail, 
-                                                              new_fl_stail) ) {
+                        if ( 0 == atomic_load(&(type_info_ptr->serial_num)) ) {
```

Comment:
Same as for ID, this should probably be checked and cause an assertion failure  right after the head is retrieved